File: app/src/main/java/com/example/xingliansdk/ui/setting/BpSettingActivity.java
package com.example.xingliansdk.ui.setting;

import android.os.Bundle;

import com.example.xingliansdk.R;
import com.example.xingliansdk.base.BaseActivity;
import com.example.xingliansdk.base.viewmodel.BaseViewModel;
import com.example.xingliansdk.ui.login.viewMode.UserViewModel;
import com.example.xingliansdk.utils.HawkUtil;
import com.example.xingliansdk.view.DateUtil;
import com.example.xingliansdk.widget.TitleBarLayout;
import com.github.iielse.switchbutton.SwitchView;
import com.gyf.barlibrary.ImmersionBar;
import com.luck.picture.lib.tools.ToastUtils;
import com.shon.HawConstant;
import com.shon.connector.BleWrite;
import com.shon.connector.bean.AutoBpStatusBean;
import com.shon.connector.call.listener.CommBackListener;
import com.shon.connector.utils.ShowToast;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import androidx.annotation.Nullable;

/**
 * 血压自动测量状态设置页面
 * Created by Admin
 * Date 2022/5/7
 */
public class BpSettingActivity extends BaseActivity<UserViewModel> {

    private TitleBarLayout titleBar;

    //夜间测量
    private SwitchView autoBpNightSwitch;
    //正常测量
    private SwitchView autoBpNormalSwitch;

    private AutoBpStatusBean autoBpStatusBean;

    @Override
    public int layoutId() {
        return R.layout.activity_bp_set_layout;
    }

    @Override
    public void initView(@Nullable Bundle savedInstanceState) {
        findViews();
        ImmersionBar.with(this)
                .titleBar(titleBar)
                .init();


        titleBar.setTitleBarListener(new TitleBarLayout.TitleBarListener() {
            @Override
            public void onBackClick() {
                finish();
            }

            @Override
            public void onActionImageClick() {

            }

            @Override
            public void onActionClick() {
                setBpStatus();
            }
        });
    }

    private void findViews(){
        titleBar = findViewById(R.id.titleBar);
        autoBpNightSwitch = findViewById(R.id.autoBpNightSwitch);
        autoBpNormalSwitch = findViewById(R.id.autoBpNormalSwitch);

        autoBpNormalSwitch.setOnStateChangedListener(onStateChangedListener);
        autoBpNightSwitch.setOnStateChangedListener(onStateChangedListener);

        initData();
    }


    @Override
    public void createObserver() {
        super.createObserver();



    }

    private void initData(){
        if(autoBpStatusBean == null)
         autoBpStatusBean = new AutoBpStatusBean();

        List<Boolean> saveList = HawConstant.getAutoBpStatus();
        if(saveList != null){
            autoBpNightSwitch.setOpened(saveList.get(1));
            autoBpNormalSwitch.setOpened(saveList.get(0));
        }

    }

    private final SwitchView.OnStateChangedListener onStateChangedListener = new SwitchView.OnStateChangedListener() {
        @Override
        public void toggleToOn(SwitchView view) {
            if(view.getId() == R.id.autoBpNightSwitch){ //夜间
                autoBpNightSwitch.setOpened(true);
                autoBpStatusBean.setNightBpStatus((byte) 0x02);
            }
            if(view.getId() == R.id.autoBpNormalSwitch){    //正常
                autoBpNormalSwitch.setOpened(true);
               autoBpStatusBean.setNormalBpStatus((byte) 0x02);
            }
        }

        @Override
        public void toggleToOff(SwitchView view) {
            if(view.getId() == R.id.autoBpNightSwitch){ //夜间
                autoBpNightSwitch.setOpened(false);
                autoBpStatusBean.setNightBpStatus((byte) 0x01);
            }
            if(view.getId() == R.id.autoBpNormalSwitch){    //正常
                autoBpNormalSwitch.setOpened(false);
                autoBpStatusBean.setNormalBpStatus((byte) 0x01);
            }
        }
    };


    private void setBpStatus(){
        if(autoBpNormalSwitch.isOpened()){
            autoBpStatusBean.setNormalBpStatus((byte) 0x02);
            autoBpStatusBean.setStartHour(0x08);
            autoBpStatusBean.setStartMinute(0x00);
            autoBpStatusBean.setEndHour(0x00);
            autoBpStatusBean.setEndMinute(0x00);
            autoBpStatusBean.setBpInterval(0x05);
        }


        //夜间
        if(autoBpNightSwitch.isOpened()){
            autoBpStatusBean.setNightBpStatus((byte) 0x02);
        }

        BleWrite.writeSetAutoBpMeasureStatus(true, autoBpStatusBean, new CommBackListener() {
            @Override
            public void commWriteBackData(Object object) {
                if((Boolean) object){
                    List<Boolean> booleanList = new ArrayList<>();
                    booleanList.add(autoBpNormalSwitch.isOpened());
                    booleanList.add(autoBpNightSwitch.isOpened());
                    HawConstant.saveAutoBpStatus(booleanList);
                    ShowToast.INSTANCE.showToast(BpSettingActivity.this,"设置成功",2);
                }
            }
        });


        HashMap<String,String> value = new HashMap<String, String>();
        value.put("nickname",getMDeviceInformationBean().getName());
        value.put("height",getMDeviceInformationBean().getHeight()+"");
        value.put("weight",getMDeviceInformationBean().getWeight()+"");
        value.put("createTime",(System.currentTimeMillis()/1000)+"");
        value.put("age",getMDeviceInformationBean().getAge()+"");
        value.put("sex",getMDeviceInformationBean().getSex()+"");
        value.put("birthDate", DateUtil.getDate(DateUtil.YYYY_MM_DD, getMDeviceInformationBean().getBirth()));
        value.put("blood_pressure_night_sleep_measurement",autoBpNightSwitch.isOpened() ? "1" : "2");
        value.put("blood_pressure_non_sleep_measurement",autoBpNormalSwitch.isOpened() ? "1" : "2");
        mViewModel.setUserInfo(value);

    }
}

File: app/src/main/java/com/example/xingliansdk/ui/bp/BpSettingActivity.kt
package com.example.xingliansdk.ui.bp

import android.os.Bundle
import com.example.xingliansdk.R
import com.example.xingliansdk.base.BaseActivity
import com.example.xingliansdk.network.api.login.LoginBean
import com.example.xingliansdk.ui.login.viewMode.UserViewModel
import com.example.xingliansdk.view.DateUtil
import com.example.xingliansdk.widget.TitleBarLayout.TitleBarListener
import com.github.iielse.switchbutton.SwitchView
import com.gyf.barlibrary.ImmersionBar
import com.orhanobut.hawk.Hawk
import com.shon.HawConstant
import com.shon.connector.BleWrite
import com.shon.connector.bean.AutoBpStatusBean
import com.shon.connector.utils.ShowToast
import kotlinx.android.synthetic.main.activity_bp_set_layout.*
import java.util.ArrayList
import java.util.HashMap

/**
 * Created by Admin
 *Date 2022/5/30
 */
class BpSettingActivity : BaseActivity<UserViewModel>() {


    private var autoBpStatusBean: AutoBpStatusBean? = null

    override fun layoutId(): Int {
        return R.layout.activity_bp_set_layout
    }

    override fun initView(savedInstanceState: Bundle?) {
        ImmersionBar.with(this)
            .titleBar(titleBar)
            .init()


        autoBpNormalSwitch.setOnStateChangedListener(onStateChangedListener)
        autoBpNightSwitch.setOnStateChangedListener(onStateChangedListener)

        titleBar.setTitleBarListener(object : TitleBarListener {
            override fun onBackClick() {
                finish()
            }

            override fun onActionImageClick() {}
            override fun onActionClick() {
                setBpStatus()
            }
        })

        initData()
    }

    override fun createObserver() {
        super.createObserver()
        mViewModel.result.observe(this){
            hideWaitDialog()
            ShowToast.showToastShort(resources.getString(R.string.string_save_successed))
            finish()
        }

        mViewModel.msg.observe(this){
            hideWaitDialog()
            ShowToast.showToastShort(resources.getString(R.string.string_save_failed))
        }

    }


    private fun initData() {
        if (autoBpStatusBean == null) autoBpStatusBean = AutoBpStatusBean()


        val saveBp = HawConstant.getAutoBpStatusData()

        val userInfo = Hawk.get(com.example.xingliansdk.Config.database.USER_INFO, LoginBean())
        if(userInfo != null){
            val isNormal = userInfo.userConfig.bloodPressureDaytimeMeasurement
            val isNight = userInfo.userConfig.bloodPressureNightMeasurement
            autoBpNightSwitch.isOpened = isNight == 0x02
            autoBpNormalSwitch.isOpened =isNormal == 0x02
        }else{
            autoBpNightSwitch.isOpened = false
            autoBpNormalSwitch.isOpened = false
        }
    }

    private val onStateChangedListener: SwitchView.OnStateChangedListener =
        object : SwitchView.OnStateChangedListener {
            override fun toggleToOn(view: SwitchView) {
                if (view.id == R.id.autoBpNightSwitch) { //夜间
                    autoBpNightSwitch.isOpened = true
                    autoBpStatusBean?.nightBpStatus = 0x02.toByte()
                }
                if (view.id == R.id.autoBpNormalSwitch) {    //正常
                    autoBpNormalSwitch.isOpened = true
                    autoBpStatusBean?.normalBpStatus = 0x02.toByte()
                }
            }

            override fun toggleToOff(view: SwitchView) {
                if (view.id == R.id.autoBpNightSwitch) { //夜间
                    autoBpNightSwitch.isOpened = false
                    autoBpStatusBean?.nightBpStatus = 0x01.toByte()
                }
                if (view.id == R.id.autoBpNormalSwitch) {    //正常
                    autoBpNormalSwitch.isOpened = false
                    autoBpStatusBean?.normalBpStatus = 0x01.toByte()
                }
            }
        }


    private fun setBpStatus() {
        if (autoBpNormalSwitch.isOpened) {
            autoBpStatusBean?.setNormalBpStatus(0x02.toByte())

        }else{
            autoBpStatusBean?.setNormalBpStatus(0x01.toByte())
        }
        autoBpStatusBean?.setStartHour(0x08)
        autoBpStatusBean?.setStartMinute(0x00)
        autoBpStatusBean?.setEndHour(0x17)
        autoBpStatusBean?.setEndMinute(0x3B)
        autoBpStatusBean?.setBpInterval(0x0A)

        //夜间
        autoBpStatusBean?.setNightBpStatus(if(autoBpNightSwitch.isOpened) 0x02.toByte() else 0x01)
        HawConstant.saveAutoBpData(autoBpStatusBean)
        showLoading("Loading...")
        BleWrite.writeSetAutoBpMeasureStatus(
            true, autoBpStatusBean
        ) { `object` ->
            if (`object` as Boolean) {
                val booleanList: MutableList<Boolean> =
                    ArrayList()
                booleanList.add(autoBpNormalSwitch.isOpened)
                booleanList.add(autoBpNightSwitch.isOpened)
                HawConstant.saveAutoBpStatus(booleanList)
                //                    ShowToast.INSTANCE.showToast(BpSettingActivity.this,"设置成功",2);
            }
        }
        val value = HashMap<String, String>()
        value["nickname"] = mDeviceInformationBean.name
        value["height"] = mDeviceInformationBean.height.toString() + ""
        value["weight"] = mDeviceInformationBean.weight.toString() + ""
        value["createTime"] = (System.currentTimeMillis() / 1000).toString() + ""
        value["age"] = mDeviceInformationBean.age.toString() + ""
        value["sex"] = mDeviceInformationBean.sex.toString() + ""
        value["birthDate"] = DateUtil.getDate(DateUtil.YYYY_MM_DD, mDeviceInformationBean.birth)

        value["bloodPressureNightMeasurement"] = if (autoBpNightSwitch.isOpened) "2" else "1"
        value["bloodPressureDaytimeMeasurement"] = if (autoBpNormalSwitch.isOpened) "2" else "1"
        mViewModel.setUserInfo(value)
    }
}